mesonbuild/wrap/wrap.py

```diff
@@ -312,41 +288,48 @@ def get_data(self, url):
             hashvalue = h.hexdigest()
         return hashvalue, tmpfile.name
 
-    def get_hash(self, data):
+    def check_hash(self, p, what, path):
+        expected = p.get(what + '_hash')
         h = hashlib.sha256()
-        h.update(data)
-        hashvalue = h.hexdigest()
-        return hashvalue
-
-    def download(self, p, packagename):
-        ofname = os.path.join(self.cachedir, p.get('source_filename'))
-        if os.path.exists(ofname):
-            mlog.log('Using', mlog.bold(packagename), 'from cache.')
-        else:
-            srcurl = p.get('source_url')
-            mlog.log('Downloading', mlog.bold(packagename), 'from', mlog.bold(srcurl))
-            dhash, tmpfile = self.get_data(srcurl)
-            expected = p.get('source_hash')
-            if dhash != expected:
-                os.remove(tmpfile)
-                raise RuntimeError('Incorrect hash for source %s:\n %s expected\n %s actual.' % (packagename, expected, dhash))
-            os.rename(tmpfile, ofname)
-        if p.has_patch():
-            patch_filename = p.get('patch_filename')
-            filename = os.path.join(self.cachedir, patch_filename)
-            if os.path.exists(filename):
-                mlog.log('Using', mlog.bold(patch_filename), 'from cache.')
-            else:
-                purl = p.get('patch_url')
-                mlog.log('Downloading patch from', mlog.bold(purl))
-                phash, tmpfile = self.get_data(purl)
-                expected = p.get('patch_hash')
-                if phash != expected:
-                    os.remove(tmpfile)
-                    raise RuntimeError('Incorrect hash for patch %s:\n %s expected\n %s actual' % (packagename, expected, phash))
-                os.rename(tmpfile, filename)
-        else:
-            mlog.log('Package does not require patch.')
+        with open(path, 'rb') as f:
+            h.update(f.read())
+        dhash = h.hexdigest()
+        if dhash != expected:
+            raise WrapException('Incorrect hash for %s:\n %s expected\n %s actual.' % (what, expected, dhash))
+
+    def download(self, p, what, ofname):
+        self.check_can_download()
+        srcurl = p.get(what + '_url')
+        mlog.log('Downloading', mlog.bold(self.packagename), what, 'from', mlog.bold(srcurl))
+        dhash, tmpfile = self.get_data(srcurl)
+        expected = p.get(what + '_hash')
+        if dhash != expected:
+            os.remove(tmpfile)
+            raise WrapException('Incorrect hash for %s:\n %s expected\n %s actual.' % (what, expected, dhash))
+        os.rename(tmpfile, ofname)
+
+    def get_file_internal(self, p, what):
+        filename = p.get(what + '_filename')
+        cache_path = os.path.join(self.cachedir, filename)
+
+        if os.path.exists(cache_path):
+            self.check_hash(p, what, cache_path)
+            mlog.log('Using', mlog.bold(self.packagename), what, 'from cache.')
+            return cache_path
+
```

Comment:
We need a `self.check_can_download()` here to allow for cached files but disallow downloading.

Reply:
The check is already in the download() method.

Reply:
Yes, i missed that somehow.